File: talentflow/tools/extraction_preview.py
"""Extraction preview: per-section confidence check that runs before scoring.

Mirrors web/src/utils/extractionPreview.ts so the CLI agent loop and the web
UI apply the same rule: header/skills/experience/projects/education each get
a confidence score, and anything missing or malformed is flagged before
evaluate_fit ever sees the resume.
"""

from __future__ import annotations

import re
from typing import Any

_REQUIRED_SECTIONS = ("skills", "experience", "education")

_LABELS = {
    "header": "Header",
    "skills": "Skills",
    "experience": "Experience",
    "projects": "Projects",
    "education": "Education",
}

_SECTION_LOOKS_LIKE_BODY = re.compile(r"^(skills?|experience|education|projects?)\s*:", re.I)
_PROJECTS_HEADING = re.compile(r"\bprojects?\b\s*:?", re.I)


def _clamp01(value: float) -> float:
    return round(min(1.0, max(0.0, value)), 2)


def _header_preview(raw_text: str) -> dict[str, Any]:
    first_line = next((line.strip() for line in raw_text.splitlines() if line.strip()), "")
    detected = bool(first_line) and not _SECTION_LOOKS_LIKE_BODY.match(first_line)
    confidence = _clamp01(0.85 if detected and len(first_line) > 15 else 0.5) if detected else 0.0

    return {
        "key": "header",
        "label": _LABELS["header"],
        "detected": detected,
        "confidence": confidence,
        "required": False,
        "summary": first_line[:120] if detected else "No header/contact line detected",
    }


def _skills_preview(skills: list[str]) -> dict[str, Any]:
    count = len(skills)
    if count >= 6:
        confidence = 0.92
    elif count >= 3:
        confidence = 0.75
    elif count >= 1:
        confidence = 0.5
    else:
        confidence = 0.0

    detected = count > 0
    warning = None
    if not detected:
        warning = f"{_LABELS['skills']} section not detected."
    elif confidence < 0.6:
        warning = f"{_LABELS['skills']} section low confidence."

    return {
        "key": "skills",
        "label": _LABELS["skills"],
        "detected": detected,
        "confidence": _clamp01(confidence),
        "required": True,
        "summary": f"{count} skill(s) found" if detected else "No skills detected",
        "warning": warning,
    }


def _experience_preview(experience: list[dict[str, Any]]) -> dict[str, Any]:
    count = len(experience)
    complete = sum(
        1
        for item in experience
        if item.get("title") and item.get("company") and item.get("summary") and item.get("years", 0) > 0
    )

    confidence = 0.0
    if count > 0:
        confidence = 0.5 + 0.15 * min(count, 3)
        if complete < count:
            confidence -= 0.12

    detected = count > 0
    warning = None
    if not detected:
        warning = f"{_LABELS['experience']} section not detected."
    elif confidence < 0.6:
        warning = f"{_LABELS['experience']} section low confidence."

    return {
        "key": "experience",
        "label": _LABELS["experience"],
        "detected": detected,
        "confidence": _clamp01(confidence),
        "required": True,
        "summary": f"{count} entry(ies) found" if detected else "No experience detected",
        "warning": warning,
    }


def _education_preview(education: list[dict[str, Any]]) -> dict[str, Any]:
    count = len(education)
    complete = sum(1 for item in education if item.get("degree") and item.get("school"))

    detected = count > 0
    confidence = 0.0
    if detected:
        confidence = 0.85 if complete == count else 0.55

    warning = None
    if not detected:
        warning = f"{_LABELS['education']} section not detected."
    elif confidence < 0.6:
        warning = f"{_LABELS['education']} section low confidence."

    return {
        "key": "education",
        "label": _LABELS["education"],
        "detected": detected,
        "confidence": _clamp01(confidence),
        "required": True,
        "summary": f"{count} entry(ies) found" if detected else "No education detected",
        "warning": warning,
    }


def _projects_preview(raw_text: str) -> dict[str, Any]:
    match = _PROJECTS_HEADING.search(raw_text)
    detected = bool(match)
    content_len = len(raw_text[match.end():]) if match else 0
    confidence = _clamp01(0.78 if content_len > 40 else 0.55) if detected else 0.0

    return {
        "key": "projects",
        "label": _LABELS["projects"],
        "detected": detected,
        "confidence": confidence,
        "required": False,
        "summary": "Projects section found" if detected else "No projects section (optional)",
    }


def build_extraction_preview(resume: dict[str, Any]) -> dict[str, Any]:
    """Compute a per-section confidence preview for a resume before scoring.

    Uses the already-structured skills/experience/education lists (reliable)
    plus a light scan of raw_text for header/projects (which have no
    structured field). Only skills/experience/education gate validity —
    header/projects are informational.
    """
    raw_text = resume.get("raw_text") or ""

    sections = [
        _header_preview(raw_text),
        _skills_preview(resume.get("skills", [])),
        _experience_preview(resume.get("experience", [])),
        _education_preview(resume.get("education", [])),
        _projects_preview(raw_text),
    ]

    warnings = [section["warning"] for section in sections if section.get("warning")]

    required_detected = sum(
        1 for section in sections if section["key"] in _REQUIRED_SECTIONS and section["detected"]
    )
    if required_detected == 0:
        warnings.append("Resume format unclear.")

    is_valid = all(
        next(s for s in sections if s["key"] == key)["detected"] for key in _REQUIRED_SECTIONS
    )

    return {"sections": sections, "warnings": warnings, "is_valid": is_valid}


def format_extraction_preview(candidate_name: str, preview: dict[str, Any]) -> list[str]:
    """Render the preview as CLI-friendly lines, one block per section."""
    status = "validated" if preview["is_valid"] else "needs review"
    lines = [f"{candidate_name} — extraction preview ({status})"]

    for section in preview["sections"]:
        pct = round(section["confidence"] * 100)
        tag = "" if section["required"] else " (optional)"
        lines.append(f"  - {section['label']}{tag}: {pct}% — {section['summary']}")

    for warning in preview["warnings"]:
        lines.append(f"  ⚠ {warning}")

    return lines

File: web/src/utils/extractionPreview.ts
import type { ResumeData } from "../types";
import { parseSections, type ParsedSection } from "./skillParser";

export type PreviewSectionKey = "header" | "skills" | "experience" | "projects" | "education";

export interface SectionPreview {
  key: PreviewSectionKey;
  label: string;
  detected: boolean;
  confidence: number;
  required: boolean;
  summary: string;
  warning?: string;
}

export interface ExtractionPreviewResult {
  sections: SectionPreview[];
  warnings: string[];
  isValid: boolean;
}

const LABELS: Record<PreviewSectionKey, string> = {
  header: "Header",
  skills: "Skills",
  experience: "Experience",
  projects: "Projects",
  education: "Education",
};

/** Only these blocks gate scoring — Header/Projects are informational, not required. */
const REQUIRED_SECTIONS: PreviewSectionKey[] = ["skills", "experience", "education"];

function clamp01(value: number): number {
  return Math.min(1, Math.max(0, Math.round(value * 100) / 100));
}

function countTokens(text: string): number {
  return text
    .split(/[,;|•\n]/)
    .map((token) => token.trim())
    .filter((token) => token.length > 1).length;
}

function findParsedSection(sections: ParsedSection[], name: ParsedSection["name"]): string {
  return sections
    .filter((section) => section.name === name)
    .map((section) => section.content)
    .filter(Boolean)
    .join("\n");
}

function headerContent(sections: ParsedSection[], rawText: string): string {
  const first = sections[0];
  if (first && first.name === "unknown") {
    return [first.heading === "Preamble" ? "" : first.heading, first.content]
      .filter(Boolean)
      .join(" ")
      .trim();
  }
  const firstLine = rawText.split(/\r?\n/).find((line) => line.trim().length > 0) ?? "";
  return firstLine.trim();
}

function previewHeader(sections: ParsedSection[], rawText: string): SectionPreview {
  const content = headerContent(sections, rawText);
  const looksLikeSectionBody = /^(skills?|experience|education|projects?)\s*:/i.test(content);
  const detected = content.length > 0 && !looksLikeSectionBody;
  const confidence = detected ? clamp01(content.length > 15 ? 0.85 : 0.5) : 0;

  return {
    key: "header",
    label: LABELS.header,
    detected,
    confidence,
    required: false,
    summary: detected ? content.slice(0, 120) : "No header/contact line detected",
  };
}

function previewSkills(resume: ResumeData, sections: ParsedSection[]): SectionPreview {
  const structuredCount = resume.skills?.length ?? 0;
  const parsedTokenCount = countTokens(findParsedSection(sections, "skills"));
  const count = Math.max(structuredCount, parsedTokenCount);

  let confidence = 0;
  if (count >= 6) confidence = 0.92;
  else if (count >= 3) confidence = 0.75;
  else if (count >= 1) confidence = 0.5;

  const detected = count > 0;
  const warning = !detected
    ? `${LABELS.skills} section not detected.`
    : confidence < 0.6
      ? `${LABELS.skills} section low confidence.`
      : undefined;

  return {
    key: "skills",
    label: LABELS.skills,
    detected,
    confidence: clamp01(confidence),
    required: true,
    summary: detected ? `${count} skill${count === 1 ? "" : "s"} found` : "No skills detected",
    warning,
  };
}

function previewExperience(resume: ResumeData): SectionPreview {
  const entries = resume.experience ?? [];
  const count = entries.length;
  const complete = entries.filter(
    (item) => item.title && item.company && item.summary && item.years > 0,
  ).length;

  let confidence = 0;
  if (count > 0) {
    confidence = 0.5 + 0.15 * Math.min(count, 3);
    if (complete < count) confidence -= 0.12;
  }

  const detected = count > 0;
  const warning = !detected
    ? `${LABELS.experience} section not detected.`
    : confidence < 0.6
      ? `${LABELS.experience} section low confidence.`
      : undefined;

  return {
    key: "experience",
    label: LABELS.experience,
    detected,
    confidence: clamp01(confidence),
    required: true,
    summary: detected ? `${count} entr${count === 1 ? "y" : "ies"} found` : "No experience detected",
    warning,
  };
}

function previewEducation(resume: ResumeData): SectionPreview {
  const entries = resume.education ?? [];
  const count = entries.length;
  const complete = entries.filter((item) => item.degree && item.school).length;

  const detected = count > 0;
  const confidence = detected ? (complete === count ? 0.85 : 0.55) : 0;
  const warning = !detected
    ? `${LABELS.education} section not detected.`
    : confidence < 0.6
      ? `${LABELS.education} section low confidence.`
      : undefined;

  return {
    key: "education",
    label: LABELS.education,
    detected,
    confidence: clamp01(confidence),
    required: true,
    summary: detected ? `${count} entr${count === 1 ? "y" : "ies"} found` : "No education detected",
    warning,
  };
}

function previewProjects(sections: ParsedSection[]): SectionPreview {
  const content = findParsedSection(sections, "projects");
  const detected = content.trim().length > 0;
  const confidence = detected ? clamp01(content.length > 40 ? 0.78 : 0.55) : 0;

  return {
    key: "projects",
    label: LABELS.projects,
    detected,
    confidence,
    required: false,
    summary: detected ? "Projects section found" : "No projects section (optional)",
  };
}

/**
 * Build the extraction_preview step output: one block per resume section with a
 * confidence score, plus warnings for anything missing or malformed. This runs
 * after extraction and before skill parsing/scoring so a human can catch a bad
 * parse before it silently skews the fit score.
 */
export function buildExtractionPreview(resume: ResumeData): ExtractionPreviewResult {
  const rawText = resume.raw_text ?? "";
  const sections = resume.raw_text ? parseSections(resume.raw_text) : [];

  const previewSections: SectionPreview[] = [
    previewHeader(sections, rawText),
    previewSkills(resume, sections),
    previewExperience(resume),
    previewEducation(resume),
    previewProjects(sections),
  ];

  const warnings = previewSections
    .map((section) => section.warning)
    .filter((warning): warning is string => Boolean(warning));

  const requiredDetected = previewSections.filter(
    (section) => REQUIRED_SECTIONS.includes(section.key) && section.detected,
  ).length;

  if (requiredDetected === 0) {
    warnings.push("Resume format unclear.");
  }

  const isValid = REQUIRED_SECTIONS.every((key) =>
    previewSections.find((section) => section.key === key)?.detected,
  );

  return { sections: previewSections, warnings, isValid };
}
